fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java

```diff
@@ -0,0 +1,324 @@
+            Preconditions.checkNotNull(srcReplica);
+            TabletScheduler.PathSlot slot = backendsWorkingSlots.get(srcReplica.getBackendId());
+            Preconditions.checkNotNull(slot, "unable to get fromBe " + srcReplica.getBackendId() + " slot");
+            if (slot.takeBalanceSlot(srcReplica.getPathHash()) != -1) {
+                tabletCtx.setSrc(srcReplica);
+            } else {
+                throw new SchedException(SchedException.Status.SCHEDULE_FAILED, "no slot for src replica " + srcReplica + ", pathHash " + srcReplica.getPathHash());
+            }
+
+            // Choose a path in destination
+            ClusterLoadStatistic clusterStat = statisticMap.get(tabletCtx.getCluster());
+            Preconditions.checkNotNull(clusterStat, "cluster does not exist: " + tabletCtx.getCluster());
+            BackendLoadStatistic beStat = clusterStat.getBackendLoadStatistic(move.toBe);
+            Preconditions.checkNotNull(beStat);
+            slot = backendsWorkingSlots.get(move.toBe);
+            Preconditions.checkNotNull(slot, "unable to get slot of toBe " + move.toBe);
+
+            List<RootPathLoadStatistic> paths = beStat.getPathStatistics();
+            Set<Long> availPath = paths.stream().filter(path -> path.getStorageMedium() == tabletCtx.getStorageMedium()
+                    && path.isFit(tabletCtx.getTabletSize(), false) == BalanceStatus.OK)
+                    .map(RootPathLoadStatistic::getPathHash).collect(Collectors.toSet());
+            long pathHash = slot.takeAnAvailBalanceSlotFrom(availPath);
+            if (pathHash == -1) {
+                throw new SchedException(SchedException.Status.SCHEDULE_FAILED, "paths has no available balance slot: " + availPath);
+            } else {
+                tabletCtx.setDest(beStat.getBeId(), pathHash);
+            }
+
+            // ToDeleteReplica is the source replica
+            pair.second = srcReplica.getId();
+        } catch (IllegalStateException | NullPointerException e) {
+            // Problematic move should be invalidated immediately
+            movesInProgress.get().invalidate(tabletCtx.getTabletId());
+            throw new SchedException(SchedException.Status.UNRECOVERABLE, e.getMessage());
+        }
+    }
+
+    // The validation check cannot be accurate, cuz the production of moves do have ordering.
+    // If some moves failed, the cluster & partition skew is different to the skew when we getNextMove.
+    // So we can't do skew check.
+    // Just do some basic checks, e.g. server available.
+    private void checkMoveValidation(ReplicaMove move) throws IllegalStateException {
+        boolean fromAvailable = infoService.checkBackendAvailable(move.fromBe);
+        boolean toAvailable = infoService.checkBackendAvailable(move.toBe);
+        Preconditions.checkState(fromAvailable && toAvailable, move + "'s bes are not all available: from " + fromAvailable + ", to " + toAvailable);
+        // To be improved
+    }
```

Comment:
selectAlternativeTabletsForCluster method selects all tablets for maxSkew partition from highload be. If there is no check whether the move will make partition balance better, we may fall into endless loop in such case:
3 tablets on maxload be
1 tablet on minload be
we move 2 tablets from maxload be to minload be, and maxload be switchs to minload be.

Reply:
Not all, just pick one tablet to generate one move. and apply the move to current balance info, it's in the algo implement.

Reply:
OK